### ISSUE TITLE
An option for disabling activation of autoloads globally and (maybe) per package

### ISSUE THREAD

**Issue by piotrklibert** [radian-software/straight.el#914]
Using the `straight-cache-autoloads` helps a bit, but loading - reading and evaling - hundreds of packages' autoloads (currently 350 in my config) slows down the startup significantly (10%-20% in my crude testing). It's especially irritating because `use-package` already sets up the autoloads, plus the vast majority of the packages won't be loaded at all in a given session. I'm also perfectly fine with putting `(require 'X-autoloads)` where it's really needed myself.

The unconditional load of package autoloads seems to happen here:  https://github.com/raxod502/straight.el/blob/master/straight.el#L5761

I would like to introduce a global setting variable that would control this call. Disabling this per package would be harder, and honestly, I'm not sure there's a use case for this, so I'm only mentioning this for completeness - I think most other settings can be overridden either by args to `straight-use-package` or in the recipe plist.

Basically, I would like to be able to tell straight that I'm happy it built the loaddefs files for me, thanks, but I will take it from there. I want to manage them myself. I'm not sure it will help anyone other than myself, and I'm also not sure if anything else would break due to this, even though the change seems relatively minor. I'm opening an issue to address this uncertainty, and also to ask if the idea somehow conflicts with straight goals/assumptions. If there's nothing wrong with adding it, I can make a PR with it.

**Comment by progfolio** [comment]
> slows down the startup significantly (10%-20% in my crude testing)

Would you mind sharing the tests and results?

**Comment by progfolio** [comment]
>  It's especially irritating because use-package already sets up the autoloads

Straight.el can be used without use-package, so we can't make that assumption on our end.
I believe the use-package macro also only conditionally autoloads if necessary when certain keywords are used
(:bind, :commands, :mode, :interpreter, etc). e.g.


```emacs-lisp
(use-package test :commands (command))
```

Expands to:

```emacs-lisp
(progn
  (straight-use-package 'test)
  (unless
      (fboundp 'command)
    (autoload
      (function command)
      "test" nil t)))
```

The calls to `fboundp` should be immaterial. For example on my system I get the following:


```emacs-lisp
(benchmark 30000 '(fboundp 'unbound)) ;"Elapsed time: 0.003263s"
(benchmark 30000 '(fboundp 'ignore))  ;"Elapsed time: 0.003645s"
```

So the time cost is most likely not in the duplication of autoload effort in those use-package cases.


Anecdonatally, ignoring autoload activation on my system by advising `straight--activate-package-autoloads`:


```emacs-lisp
(advice-add 'straight--activate-package-autoloads :around #'ignore)
```

Is about 0.05 seconds faster (roughly a 3-4% reduction in start-up time). 
This is with roughly 110 packages. Assuming this scales linearly, are we talking a differnce of about 0.12 to 0.25 seconds in the scenario you described with 300 packages?

> I'm also perfectly fine with putting (require 'X-autoloads) where it's really needed myself.
> [...]
> I'm not sure there's a use case for this.

I'm not sure either. I would imagine there are more users who benefit from the autoloads than those who suffer a significant startup time penalty due to activating them. In either case, I think it would be better to try to improve the time it takes to activate the autoloads first.

**Comment by progfolio** [comment]
I've pushed a change that may help on my personal fork. Would  you mind testing it out to see if it improves the performance for you at all?

You can test it by adding the following prior to the straight.el bootstrap snippet in your init file:

```emacs-lip
(setq straight-repository-user "progfolio"
         straight-repository-branch "maybe/faster-autoload-activation")
```

**Comment by piotrklibert** [comment]
I will test it next week, thank you for a quick response. I don't have the benchmark results on hand right now, but from memory: with 350 active packages the autoloads took 0.4-0.6 sec on my machine. This might be specific to my environment, ie. I might have a package installed that has autoloads that misbehave, increasing the load time more than expected. Apparently, packages are free to put *any code* into their autoloads, so I think it's possible.

About `use-package` - yes, you're right, the `:commands` part expands to essentially no-ops when a function of a given name is already defined (no matter if it's an autoload object or actual function object). But, in that case, what is the use for `:commands` keyword? The autoloads are already *all* loaded; you can delete the whole `:commands` sections from all the `use-package` invocations and nothing will change. That's more of a problem with `use-package` design, though. I'm investigating leaf as an alternative right now.

In any case - thanks for the response and let's continue this discussion in the New Year. Hopefully, it'll also be Happier, for all of us. :)

**Comment by progfolio** [comment]
> what is the use for :commands keyword?

It can be used in cases where the package does not provide proper autoloads.
It also implicitly defers loading the package when used in combination with other keywords (e.g. `:config`).

```emacs-lisp
(use-package test
  :config (ignore))
```
Which expands to:

```emacs-lisp
(progn
  (straight-use-package 'test)
  (require 'test nil nil)
  (ignore)
  t)
```

And:

```emacs-lisp
(use-package test
  :commands (command)
  :config (ignore))
```

Which expands to:

```emacs-lisp
(progn
  (straight-use-package 'test)
  (unless
      (fboundp 'command)
    (autoload
      (function command)
      "test" nil t))
  (eval-after-load 'test
    '(progn
       (ignore)
       t)))
```

> I will test it next week, thank you for a quick response. 

Thank you. I don't know if it will make much of a difference, but it's worth testing.

>  I might have a package installed that has autoloads that misbehave, increasing the load time more than expected. Apparently, packages are free to put any code into their autoloads, so I think it's possible.

Yes. This could be the case. It might be useful to add some advice to `straight--activate-package-autoloads` near the top of your init file to print a little report showing each package and how much time the activation took. The following should proudce a "\*Autoload Activation Timer*" buffer after start-up with activation time sorted by package. I'd be curious to see what this prints in your case.

```emacs-lisp
(let (times)
  (require 'cl-lib)
  (define-advice straight--activate-package-autoloads (:around (fn &rest args) "timer")
    (let ((start (current-time))
          (recipe (car args)))
      (apply fn args)
      (push (cons (plist-get recipe :package) (float-time (time-since start))) times)))
  (add-hook 'emacs-startup-hook
            (lambda ()
              (let ((formatter "%-25s %.06fs\n"))
                (with-current-buffer (get-buffer-create "*Autoload Activation Timer*")
                  (erase-buffer)
                  (dolist (time (cl-sort (copy-tree times) #'> :key #'cdr))
                    (insert (format formatter (car time) (cdr time))))
                  (insert (format formatter
                                  (format "Loaded %d packages in " (length times))
                                  (cl-reduce #'+ times :key #'cdr)))))
              (setq times nil))))
```
> In any case - thanks for the response and let's continue this discussion in the New Year. Hopefully, it'll also be Happier, for all of us. :)

Sounds good. Happy New Year.

**Comment by progfolio** [comment]
Any luck with the above?

**Comment by raxod502** [comment]
This thread is being closed automatically by [Tidier](https://github.com/radian-software/tidier) because it is labeled with "waiting on response" and has not seen any activity for 90 days. But don't worry—if you have any information that might advance the discussion, leave a comment and I will be happy to reopen the thread :)
